src/network_service/sending_targets_cache.rs

```diff
@@ -72,83 +56,78 @@ impl SendingTargetsCache {
-                TargetState::Failed(_) => {
-                    log_or_panic!(LogLevel::Error, "Got a failure from a failed target!");
-                }
-                TargetState::Sending(x) => {
-                    *state = TargetState::Failed(x + 1);
-                }
-                TargetState::Sent => {
-                    log_or_panic!(
-                        LogLevel::Error,
-                        "A target that should no longer fail - failed!"
-                    );
-                }
-            });
-    }
+        if entry.get().is_empty() {
+            let _ = entry.remove();
+            return None;
+        }
 
-    /// Finds a Failed target with the lowest number of failed attempts so far, among the ones that
-    /// failed at most MAX_RESENDS times. If there are multiple possibilities, the one with the
-    /// highest priority (earliest in the list) is taken. Returns None if no such targets exist.
-    fn take_next_target(&mut self, token: Token) -> Option<ConnectionInfo> {
-        self.target_states_mut(token)
-            .filter(|(_info, state)| !state.is_complete())
+        let (info, num, state) = entry
+            .get_mut()
+            .iter_mut()
             .filter_map(|(info, state)| match state {
                 TargetState::Failed(x) => Some((info, *x, state)),
-                _ => None,
-            })
-            .min_by_key(|(_info, num, _state)| *num)
-            .map(|(info, num, state)| {
-                *state = TargetState::Sending(num);
-                info
+                TargetState::Sending(_) => None,
             })
-            .cloned()
-    }
+            .min_by_key(|(_info, num, _state)| *num)?;
+        *state = TargetState::Sending(num);
 
-    fn should_drop(&self, token: Token) -> bool {
-        // Other methods maintain the invariant that exactly one of these is true:
-        // - some target is in the Sending state
-        // - we succeeded (no further sending needed)
-        // - we failed (no more targets available)
-        // So if none are sending, the handling of the message is finished and we can drop it
-        self.target_states(token)
-            .all(|(_info, state)| !state.is_sending())
+        Some(info.clone())
     }
 
-    pub fn target_failed(&mut self, token: Token, target: SocketAddr) -> Option<ConnectionInfo> {
-        self.fail_target(token, target);
-        let result = self.take_next_target(token);
-        if self.should_drop(token) {
-            let _ = self.cache.remove(&token);
+    /// Mark the given target as succeeded in sending the given message.
+    pub fn target_succeeded(&mut self, token: Token, target: SocketAddr) {
+        let mut entry = if let Entry::Occupied(entry) = self.cache.entry(token) {
+            entry
+        } else {
+            return;
+        };
+
+        entry.get_mut().retain(|(info, _)| info.peer_addr != target);
+
+        if entry.get().is_empty() {
```

Comment:
Is that correct, how does it handle send to 3 out of 7 targets like the code did before?
Not 100% sure the refactoring makes thing clearer.

Reply:
We remove only the target that succeeded sending. The ones that are still in `Sending` state are kept until they too are done sending.

Reply:
For 3 out of 7 targets (i.e hop messages), not all nodes are in Sending at the start. i.e we are suppose to complete even if some node have not been sent too.

Reply:
> Not 100% sure the refactoring makes thing clearer.

There is fewer unreachable states (so fewer log_and_panics), no dependency on implicit ordering of calls (`target_failed` and `take_next_target`) and it's slightly more efficient (fewer lookups). And the code size is slightly reduced too. Seems worthwhile to me, not sure.

Reply:
> For 3 out of 7 targets (i.e hop messages), not all nodes are in Sending at the start. i.e we are suppose to complete even if some node have not been sent too.

Ah, I see what you mean now. Yeah, that's an oversight. We should remove when there are no more `Sending`.

Reply:
Not sure the functionality is the same. Additionally, we discard information as we progress which is not always ideal (i.e when the message is failed to be sent to all its required targets, it could be good to see what are original targets, and which succeeded). while one of the `log_or_panic` was removed, its condition is not removed, i.e it would still be a bug to receive an event regarding an already sent event, we just ignore it now instead of logging it out.

Reply:
> when the message is failed to be sent to all its required targets, it could be good to see what are original targets, and which succeeded

We weren't doing that before so I didn't know that we want that. Btw, I actually added a log now which prints when a message completely fails to be sent. I doesn't print which targets failed and which succeeded, but it's better than nothing.

> one of the log_or_panic was removed, its condition is not removed

Fair enough. I'll add it back.